# Introduction

In this project I will be analysing the following [dataset](https://www.kaggle.com/datasets/shubhambathwal/flight-price-prediction) that is derived from a flight booking website.<br>

The following tasks will be performed:
1. Data cleaning using Python
2. Exploratory data analysis (EDA) using python libraries (Pandas, Matplotlib)
3. Data visualisation using Matplotlib and Seaborn

## Importing Required Libraries

In [1]:
# import the required python libraries

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

## Reading Dataset

In [2]:
# import the dataset using the first column as the index

df = pd.read_csv("C:\\Users\\Jana\\Desktop\\Flight Analysis\\flight_dataset.csv", index_col = 0)

In [3]:
# display first 5 rows

df.head()

airline   flight source_city departure_time stops   arrival_time  \
0  SpiceJet  SG-8709       Delhi        Evening  zero          Night   
1  SpiceJet  SG-8157       Delhi  Early_Morning  zero        Morning   
2   AirAsia   I5-764       Delhi  Early_Morning  zero  Early_Morning   
3   Vistara   UK-995       Delhi        Morning  zero      Afternoon   
4   Vistara   UK-963       Delhi        Morning  zero        Morning   

  destination_city    class  duration  days_left  price  
0           Mumbai  Economy      2.17          1   5953  
1           Mumbai  Economy      2.33          1   5953  
2           Mumbai  Economy      2.17          1   5956  
3           Mumbai  Economy      2.25          1   5955  
4           Mumbai  Economy      2.33          1   5955

## Initial Inspection

In [4]:
# view the dataset's dimentionality (rows, columns)

df.shape

(300153, 11)

In [5]:
# get information about the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 300153 entries, 0 to 300152
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   airline           300153 non-null  object 
 1   flight            300153 non-null  object 
 2   source_city       300153 non-null  object 
 3   departure_time    300153 non-null  object 
 4   stops             300153 non-null  object 
 5   arrival_time      300153 non-null  object 
 6   destination_city  300153 non-null  object 
 7   class             300153 non-null  object 
 8   duration          300153 non-null  float64
 9   days_left         300153 non-null  int64  
 10  price             300153 non-null  int64  
dtypes: float64(1), int64(2), object(8)
memory usage: 27.5+ MB


### About the Dataset

The dataset contains the following attributes: <br>

1. **Airline:** The name of the airline company.
2. **Flight Code:** The plane's flight code which consists of a two-character airline designator and a 1 to 4 digit number.
4. **Source City:** The departure city which is the city where the journey begins and the passenger starts their travel.
5. **Departure Period:** The departure period of the flight created by grouping time periods into bins.
6. **Stops:** The number of stops between the source and destination cities.
7. **Arrival Period:** The arrival period of the flight created by grouping time periods into bins.
8. **Destination City:** The final city that marks the end of the flight.
9. **Class:** The specific cabin or level of service depending on the seat.
10. **Duration:** The overall time it takes to travel between cities in hours.
11. **Days Left:** The number of days between booking and flight date.
12. **Price:** The ticket price provided in the same currency.

## Data Cleaning

In [6]:
# rename some columns for better understanding

df.rename(columns={'flight': 'flight_code' , 'departure_time': 'departure_period' , 'arrival_time': 'arrival_period'}, inplace = True)
df.head()

airline flight_code source_city departure_period stops arrival_period  \
0  SpiceJet     SG-8709       Delhi          Evening  zero          Night   
1  SpiceJet     SG-8157       Delhi    Early_Morning  zero        Morning   
2   AirAsia      I5-764       Delhi    Early_Morning  zero  Early_Morning   
3   Vistara      UK-995       Delhi          Morning  zero      Afternoon   
4   Vistara      UK-963       Delhi          Morning  zero        Morning   

  destination_city    class  duration  days_left  price  
0           Mumbai  Economy      2.17          1   5953  
1           Mumbai  Economy      2.33          1   5953  
2           Mumbai  Economy      2.17          1   5956  
3           Mumbai  Economy      2.25          1   5955  
4           Mumbai  Economy      2.33          1   5955

In [7]:
# count missing values

print(df.isnull().sum())

airline             0
flight_code         0
source_city         0
departure_period    0
stops               0
arrival_period      0
destination_city    0
class               0
duration            0
days_left           0
price               0
dtype: int64


**This shows that there are no _null_ or missing values in the dataframe.**

In [8]:
# detect duplicate rows

has_duplicates = df.duplicated().any()
print(f"Does the dataframe have any duplicate rows?\n{has_duplicates}")

Does the dataframe have any duplicate rows?
False


In [9]:
# remove leading and trailing spaces

df_stripped = df.map(lambda x: x.strip() if isinstance(x, str) else x)
print(df_stripped)

         airline flight_code source_city departure_period stops  \
0       SpiceJet     SG-8709       Delhi          Evening  zero   
1       SpiceJet     SG-8157       Delhi    Early_Morning  zero   
2        AirAsia      I5-764       Delhi    Early_Morning  zero   
3        Vistara      UK-995       Delhi          Morning  zero   
4        Vistara      UK-963       Delhi          Morning  zero   
...          ...         ...         ...              ...   ...   
300148   Vistara      UK-822     Chennai          Morning   one   
300149   Vistara      UK-826     Chennai        Afternoon   one   
300150   Vistara      UK-832     Chennai    Early_Morning   one   
300151   Vistara      UK-828     Chennai    Early_Morning   one   
300152   Vistara      UK-822     Chennai          Morning   one   

       arrival_period destination_city     class  duration  days_left  price  
0               Night           Mumbai   Economy      2.17          1   5953  
1             Morning           Mumba

In [10]:
# save the clean data

df2 = df_stripped
df2.to_csv('cleaned_data.csv', index = False)
print("Cleaned data saved to 'cleaned_data.csv'")

Cleaned data saved to 'cleaned_data.csv'


## Summary Statistics

In [19]:
# view statistics of numerical columns

df2.describe().T

count          mean           std      min      25%      50%  \
duration   300153.0     12.221021      7.191997     0.83     6.83    11.25   
days_left  300153.0     26.004751     13.561004     1.00    15.00    26.00   
price      300153.0  20889.660523  22697.767366  1105.00  4783.00  7425.00   

                75%        max  
duration      16.17      49.83  
days_left     38.00      49.00  
price      42521.00  123071.00

The **count** for these columns is equal to the number of rows so there are no missing values.

#### Duration:

The standard deviation **std** shows that flight lengths vary significantly.

Since the **mean** is greater than the **median**, it suggests that the distribution might be slightly ___right skewed___.

The average is being pulled up by a long tail of very long flights.

**min** (50 minutes) to **max** (50 hours) shows a wide range.

#### Days Left:

**mean** is equal to the **median**, this indicates a normal or ___symmetrical___ distribution.

The standard deviation **std** shows a moderate variability of booking times.

#### Price:

The standard deviation **std** shows extreme high variability.<br>
**std** is also greater than the **mean** which indicates a highly ___right skewed___ distribution.

**min** and **max** show that we have a large range of data which indicates _great variability_.<br>
This means the data points are widely spread out or dispersed from the **mean**/average.

The **median**/50th percentile shows that 50% of values fall below 7,425 which is much less than the **mean**/average value $ ≈ $ 20,890.<br>
The **mean**/average $ ≈ $ 20,890 is much greater than the **median**/50th percentile which shows that 50% of values fall below 7,425.<br>
This indicates that: 
- Most tickets are relatively cheap, below the **75th percentile** (42,521)
- A small number of very expensive tickets are pulling the average price up
- Most people are paying closer to the median (7,425) than the mean (20,890)

There is a large gap between the **75%** (42,521) and **max** (123,071).<br>
This also indicates a _right skewed_ distribution.

In [31]:
# view the most frequent values

df2.mode()

airline flight_code source_city departure_period stops arrival_period  \
0  Vistara      UK-706       Delhi          Morning   one          Night   

  destination_city    class  duration  days_left  price  
0           Mumbai  Economy      2.17         25  54608